AzureMonitor: support workspaces function for template variables

daniellee opened the pull request:
There is no support in the Azure Log Analytics query language for returning a list of workspaces that belong to an Azure subscription. A list of workspaces can be retrieved via the Azure Monitor API. A `workspaces()` function for template variable queries enables this. The function can take a subscription id as a parameter if it is not supplied then it will use the default subscription.

Example of a template variable query that returns a list of Log Analytics workspaces:

![image](https://user-images.githubusercontent.com/434655/77011075-5c596100-696b-11ea-90bc-e8be7446ce19.png)

Fixes #18561

daniellee committed:
azuremonitor: adds support for workspaces query macro...

...for Azure Logs template variable queries

daniellee committed:
docs: azure logs workspaces templating function

oddlittlebird commented on docs/sources/features/datasources/azuremonitor.md:
```diff
@@ -274,6 +274,40 @@ There are also some Grafana variables that can be used in Azure Log Analytics qu
+| _workspaces()_                                     | Returns a list of workspaces for the default subscription.                                             |
+| _workspaces(12345678-aaaa-bbbb-cccc-123456789aaa)_ | Returns a list of workspaces for the specified subscription (the parameter can be quoted or unquoted). |
+
+Example variable queries:
```
I recommend reformatting this bulleted list into a table.

daniellee replied:
This is what it looks like as a table:

![image](https://user-images.githubusercontent.com/434655/77012436-3e413000-696e-11ea-9a59-3e408a649198.png)

daniellee replied:
Ah, something weird with the quotes there. One sec.

daniellee replied:
It was the prettier tool we use that automatically escapes dollar signs. https://github.com/prettier/prettier/issues/6213 

I disabled prettier just for this table so that no-one else on the team accidentally adds those extra slashes back.

daniellee committed:
Update docs/sources/features/datasources/azuremonitor.md

Co-Authored-By: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

oddlittlebird commented on docs/sources/features/datasources/azuremonitor.md:
```diff
@@ -274,6 +274,40 @@ There are also some Grafana variables that can be used in Azure Log Analytics qu
+
+Any Log Analytics query that returns a list of values can be used in the `Query` field in the Variable edit view. There is also one Grafana function for Log Analytics that returns a list of workspaces.
+
+Check out the [Templating]({{< relref "../../reference/templating.md" >}}) documentation for an introduction to the templating feature and the different
```
@daniellee, I know you prefer to call it Templating, but the actual h1 title is Variables.

marefr reviewed (approved):
Code wise LGTM

daniellee committed:
docs: convert list into table

daniellee committed:
docs: fixes prettier formatting problem

Prettier adds a slash before dollar signs in markdown. Disabling it
for this table with a prettier comment.

https://prettier.io/docs/en/ignore.html

oddlittlebird reviewed (approved):
LGTM

JavierGarciaJaime commented:
Hi community,

I am running into a weird error when trying to filter the list of Log Analytics Workspaces using regex expressions with the Azure Monitor datasource.

In my case, I am able to recall all the Log Analytics list from my default Azure subscription using the **workspaces( )** on the query field, and I get the correct list, but if I want to get only the ones that start with "law" using regex and use with my variable, I could not get it.

Let me provide an example:

**workspaces()** throw the following list (I have cropped the end of each)

defaultworkspace-5cb1a0b6-...
defaultworkspace-5cb1a0b6....
law-activitylog-test...
law-activitylog-prod...
DefaultWorkspace-5cb1a0b6-65d1.....

Then I use the following regex to filter only the ones starting with "law":

Regex:
^law
/^law/

The query will always return "none". I have tried dozens of regex combinations. Is there something wrong with that Datasource or it is me? How would you get only the two workspaces starting with "law"-whatever....?

Thanks in advance.

oddlittlebird commented:
Hi @JavierGarciaJaime,

Please open a new issue or ask this question at the forums at community.grafana.com. Closed PRs are not the appropriate place for discussions like this.
